mpc85xx: add support for Freescale (NXP) P2020RDB

CHKDSK88 opened the pull request:
This PR add initial support for Freescale (NXP) P2020RDB

Hardware:
SoC: P2020 2x1GHz
DRAM: 512-1GB DDR3
2 + 4 GBE (2 separate ports and four in VSC7385)
Flash: 16MB NOR, 32MB NAND, 16MB SPI-NOR
PCIE x1 and mPCIE x1
SD Reader

Interfaces:
GBE RJ45 x6
USB2.0 x1
UART x2
I2C x2
JTAG x1
SD x1
PCIE x2 (PCIE and mPCIE)

Flash instructions:
Place sysupgrade image to 0x80000 address in NOR.
Eg (no brakelines in setenv command):

setenv 'firmware_flash tftpboot $loadaddr $firmwarefile;
 protect off $norfdtaddr +$filesize; erase $norfdtaddr +$filesize;
 cp.b $loadaddr $norfdtaddr $filesize; protect on $norfdtaddr +$filesize;
 cmp.b $loadaddr $norfdtaddr $filesize'
setenv firmwarefile firmware.bin
run firmware_flash

Boot (no brakeline in setenv command):
setenv bootcmd 'setenv bootargs root=/dev/mtdblock3 rw console=$consoledev,
$baudrate rootfstype=squashfs,jffs2 $othbootargs;
bootm $norbootaddr - $norfdtaddr'
saveenv
boot

Known issues:
-Switch is unmanaged (VSC 7385 isn't connected via SPI)
-No SD reader support


Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
mpc85xx: add support for Freescale (NXP) P2020RDB

This commit add initial support for Freescale (NXP) P2020RDB

Hardware:
SoC: P2020 2x1GHz
DRAM: 512-1GB DDR3
2 + 4 GBE (2 separate ports and four in VSC7385)
Flash: 16MB NOR, 32MB NAND, 16MB SPI-NOR
PCIE x1 and mPCIE x1
SD Reader

Interfaces:
GBE RJ45 x6
USB2.0 x1
UART x2
I2C x2
JTAG x1
SD x1
PCIE x2 (PCIE and mPCIE)

Flash instructions:
Place sysupgrade image to 0x80000 address in NOR.
Eg. (no brakelines in setenv command):

setenv 'firmware_flash tftpboot $loadaddr $firmwarefile;
 protect off $norfdtaddr +$filesize; erase $norfdtaddr +$filesize;
 cp.b $loadaddr $norfdtaddr $filesize; protect on $norfdtaddr +$filesize;
 cmp.b $loadaddr $norfdtaddr $filesize'
setenv firmwarefile firmware.bin
run firmware_flash

Boot (no brakeline in setenv command):
setenv bootcmd 'setenv bootargs root=/dev/mtdblock3 rw console=$consoledev,
$baudrate rootfstype=squashfs $othbootargs;
bootm $norfdtaddr'
saveenv
boot

Known issues:
-Switch is unmanaged (VSC 7385 is connected via eLBC, driver uses SPI)
-No SD reader support

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

chunkeey commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/fsl/p2020rdb.dts:
```diff
@@ -0,0 +1,318 @@
+/*
+ * P2020 RDB Device Tree Source
+ *
+ * Copyright 2009-2012 Freescale Semiconductor Inc.
```
It looks like this dts is mostly from the upstream p2020rdb.dts by the same name isn't it?

I think you could either make a patch with your integration changes, or use the p2020rdb.dts as a template and do your customizations on top of it in a separate new p2020rdb-owrt.dts (or with some other name). Freescale did it upstream with the "p1020rdb-pc_camp_core0.dts" that does some customization on top of the "p1020rdb-pc.dts". So you could just follow that.

CHKDSK88 replied:
It's was changed dts from upstream. I change it to patch. That allow to follow upstream changes.

chunkeey commented on target/linux/mpc85xx/files/arch/powerpc/boot/dts/fsl/p2020rdb.dts:
```diff
@@ -0,0 +1,318 @@
+			};
+			firmware@40000 {
+				reg = <0x00040000 0x00ec0000>;
+				label = "firmware";
```
Usually, you only need to specify the label OR the node-name. The sysupgrade and the mtd-utils should be smart enough to figure it out. If not something is broken.

CHKDSK88 replied:
The same name convention is used in all dts in this target. I can't change node name to "partition", because it causes "Duplicate node name" error.

mkresin replied:
> The same name convention is used in all dts in this target. I can't change node name to "partition", because it causes "Duplicate node name" error.

Which shows that you are doing it wrong :-).

Why not remove the partitions at 0x40000, 0x80000, 0x400000 and add the factory partition covering the whole space instead?

CHKDSK88 replied:
I reworked partition list. 

@mkresin 
I think about use "denx,uimage" partition spliter. But at start firmware partition is dtb area. Could You recomend method for this problem? Add offset property to spliter?

chunkeey replied:
> I think about use "denx,uimage" partition spliter. But at start firmware partition is dtb area. Could You recomend method for this problem? Add offset property to spliter? 

If you want to go that route you should check if your u-boot might already support FIT.
If not, you could still try the legacy multi-file uimage approach. In both cases the dtb (+ ramdisk if you have any) can be part of the fit/uimage. 

If you are looking for pointers for the multi-file images checkout the apm821xx. I'm using it for initramfs for the WNDR4700, MR24, MX60, MBL and for the WNDAP620/660s:
https://github.com/openwrt/openwrt/blob/master/target/linux/apm821xx/image/Makefile#L72

CHKDSK88 replied:
At this moment device won't booting from FIT image. Only imxtract both images and then regular boot work. 

```
=> imxtract 02000000 kernel@1 02400000
## Copying 'kernel@1' subimage from FIT image at 02000000 ...
crc32+ sha1+    Uncompressing part 0 ... OK
=> imxtract 02000000 fdt@1 c00000
## Copying 'fdt@1' subimage from FIT image at 02000000 ...
crc32+ sha1+    Loading part 253 ... OK
=> bootm 02400000 - c00000
WARNING: adjusting available memory to 30000000
## Booting kernel from Legacy Image at 02400000 ...
   Image Name:   POWERPC OpenWrt Linux-4.19.9
   Created:      2019-01-01  18:33:32 UTC
   Image Type:   PowerPC Linux Kernel Image (gzip compressed)
   Data Size:    2899371 Bytes = 2.8 MiB
   Load Address: 00000000
   Entry Point:  00000000
   Verifying Checksum ... OK
## Flattened Device Tree blob at 00c00000
   Booting using the fdt blob at 0x00c00000
   Uncompressing Kernel Image ... OK
   Loading Device Tree to 03ffa000, end 03ffffaf ... OK
WARNING: could not find compatible node fsl-usb2-dr: FDT_ERR_NOTFOUND.
[    0.000000] Memory CAM mapping: 256/256/256 Mb, residual: 256Mb
[    0.000000] Linux version 4.19.9 (krolik@ubuntu-VirtualBox) (gcc version 7.4.0 (OpenWrt GCC 7.4.0 r8859+56-f36bc3f9b1)) #0 SMP Tue Jan 1 15:23:09 2019
[    0.000000] Using P2020 RDB machine description
```

```
=> iminfo

## Checking Image at 02000000 ...
   FIT image found
   FIT description: POWERPC OpenWrt FIT (Flattened Image Tree)
   Created:         2019-01-01  18:33:32 UTC
    Image 0 (kernel@1)
     Description:  POWERPC OpenWrt Linux-4.19.9
     Type:         Kernel Image
     Compression:  gzip compressed
     Data Start:   0x020000ec
     Data Size:    2897673 Bytes = 2.8 MiB
     Architecture: PowerPC
     OS:           Linux
     Load Address: 0x00000000
     Entry Point:  0x00000000
     Hash algo:    crc32
     Hash value:   2b83681a
     Hash algo:    sha1
     Hash value:   0d3ea6a1d0400782a2cfa1e16ea8b18d28a50e03
    Image 1 (fdt@1)
     Description:  POWERPC OpenWrt p2020rdb device tree blob
     Type:         Flat Device Tree
     Compression:  uncompressed
     Data Start:   0x022c3934
     Data Size:    12208 Bytes = 11.9 KiB
     Architecture: PowerPC
     Hash algo:    crc32
     Hash value:   268aea01
     Hash algo:    sha1
     Hash value:   cd2de619ed19b2e32246d46c322f2a6a8374d97b
    Default Configuration: 'config@1'
    Configuration 0 (config@1)
     Description:  OpenWrt
     Kernel:       kernel@1
     FDT:          fdt@1
## Checking hash(es) for FIT Image at 02000000 ...
   Hash(es) for Image 0 (kernel@1): crc32+ sha1+
   Hash(es) for Image 1 (fdt@1): crc32+ sha1+
=> bootm
WARNING: adjusting available memory to 30000000
## Booting kernel from FIT Image at 02000000 ...
   Using 'config@1' configuration
   Trying 'kernel@1' kernel subimage
     Description:  POWERPC OpenWrt Linux-4.19.9
     Type:         Kernel Image
     Compression:  gzip compressed
     Data Start:   0x020000ec
     Data Size:    2897673 Bytes = 2.8 MiB
     Architecture: PowerPC
     OS:           Linux
     Load Address: 0x00000000
     Entry Point:  0x00000000
     Hash algo:    crc32
     Hash value:   2b83681a
     Hash algo:    sha1
     Hash value:   0d3ea6a1d0400782a2cfa1e16ea8b18d28a50e03
   Verifying Hash Integrity ... crc32+ sha1+ OK
## Flattened Device Tree from FIT Image at 02000000
   Using 'config@1' configuration
   Trying 'fdt@1' FDT blob subimage
     Description:  POWERPC OpenWrt p2020rdb device tree blob
     Type:         Flat Device Tree
     Compression:  uncompressed
     Data Start:   0x022c3934
     Data Size:    12208 Bytes = 11.9 KiB
     Architecture: PowerPC
     Hash algo:    crc32
     Hash value:   268aea01
     Hash algo:    sha1
     Hash value:   cd2de619ed19b2e32246d46c322f2a6a8374d97b
   Verifying Hash Integrity ... crc32+ sha1+ OK
   Booting using the fdt blob at 0x022c3934
   Uncompressing Kernel Image ... OK
   Loading Device Tree to 03ffa000, end 03ffffaf ... OK
WARNING: could not find compatible node fsl-usb2-dr: FDT_ERR_NOTFOUND.


```

I consider leave it as is.

chunkeey replied:
> At this moment device won't booting from FIT image

it's very close though. I've seen errors like this before. In my cases this was all due to the fdt being located at a unmappable address (either due to reserve-memory or conflict with the early kernel memory mapping) and it is possible to resolve it through either clearing the fdt_high env or setting it to a different value (for example you use 0xc00000).  

You could also try to see what happens if you enable 

CONFIG_PPC_EARLY_DEBUG


As for the "multi-file images", any luck with that?

CHKDSK88 replied:
```
[    0.649807] 3 fixed-partitions partitions found on MTD device ef000000.nor
[    0.656746] Creating 3 MTD partitions on "ef000000.nor":
[    0.662065] 0x000000000000-0x000000040000 : "NOR (RO) Vitesse-7385 Firmware"
[    0.669606] 0x000000040000-0x000000f00000 : "firmware"
[    0.675217] 2 fit-fw partitions found on MTD device firmware
[    0.680880] Creating 2 MTD partitions on "firmware":
[    0.685843] 0x000000000000-0x0000002c0000 : "kernel"
[    0.691201] 0x0000002a4664-0x000000ec0000 : "rootfs"
[    0.696557] mtd: device 3 (rootfs) set to be root filesystem
[    0.702262] 1 squashfs-split partitions found on MTD device rootfs
[    0.708443] 0x0000004c0000-0x000000ec0000 : "rootfs_data"
[    0.714246] 0x000000f00000-0x000001000000 : "u-boot"
```

Success. PR updated.

chunkeey replied:
nice :+1: . Can you tell us what exactly problem was with FIT on the first time around?

CHKDSK88 replied:
I'm not exactly sure.

chunkeey commented on target/linux/mpc85xx/p2020/config-default:
```diff
@@ -0,0 +1,31 @@
+CONFIG_GENERIC_TBSYNC=y
+CONFIG_HAVE_RCU_TABLE_FREE=y
+CONFIG_LOCK_SPIN_ON_OWNER=y
+#CONFIG_MMC=y
```
# CONFIG_MMC is not set
# CONFIG_MMC_SDHCI is not set
?

CHKDSK88 replied:
Sorry. Dumb mistake.

CHKDSK88 replied:
@chunkeey 
> Do you know what's wrong with the SD-Card reader though? Is it a chip on the pcie/usb bus or is it just missing the driver?

P2020 processor have dedicated peripherial. At this moment SD-Card not working. Freescale kernel tree (kernel 4.1) have patches, which wasn't sent to/ rejected in upstream. Also in Turris project, sd-reader was used. 
 I plan to work with this after work with VSC7385.

chunkeey replied:
It looks like parts of the driver are upstream.

I found a [fsl,p2020-esdhc and fsl,p2020-rev1-esdhc binding](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/mmc/fsl-esdhc.txt) that is serviced by a
driver under /drivers/mmc host/sdhci-of-esdhc.c that just needs
CONFIG_MMC_SDHCI_PLTFM and CONFIG_MMC_SDHCI_OF_ESDHC.

So I wonder why it isn't working out of the box? But ok, you can of course add it at any time you see fit.

chunkeey replied:
> [    1.170204] mmc0: Internal clock never stabilised.
> [    1.207150] mmc0: Internal clock never stabilised.

These messages are interesting. Could you check in /proc/device-tree if you can find the esdhc node there and see if the "clock-frequency" property file has any meaningful content (i.e. not null - hexdump can help here).

CHKDSK88 replied:
@chunkeey 
Clock have value != 0:

```
root@OpenWrt:/# hexdump /proc/device-tree/soc@ffe00000/sdhc@2e000/clock-frequenc
y
0000000 0ee6 b280
```

In u-boot everything work. I can read data from reader and check info about card.

chunkeey commented on target/linux/mpc85xx/patches-4.14/104-powerpc-mpc85xx-change-P2020RDB-dts-file-for-OpenWRT.patch:
```diff
@@ -0,0 +1,128 @@
+  * option) any later version.
+  */
+ 
++/dts-v1/;
```
Is this a bug upstream or what's going on?
The /dts-v1/; should be in the p2020si-pre.dtsi too?
Is this because of the includes of gpio.h and input.h?

CHKDSK88 replied:
@chunkeey 
>Is this because of the includes of gpio.h and input.h?

Yes. But it was reworked.

chunkeey replied:
Isn't it possible to move the gpio.h and input.h"#includes" below the /include/ p2020si-pre.dtsi ?

CHKDSK88 replied:
Is it. I will make that.

chunkeey commented on target/linux/mpc85xx/patches-4.14/104-powerpc-mpc85xx-change-P2020RDB-dts-file-for-OpenWRT.patch:
```diff
@@ -0,0 +1,128 @@
+ 
++	gpio_export {
++		compatible = "gpio-export";
++		#size-cells = <0>;
```
the #size-cells shouldn't be necessary since the child nodes have no reg properties.

Also, gpio-export has a usespace equivalent:  [gpio-switch](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/init.d/gpio_switch). The main advantage is that you don't need to carry the gpio-export patch and gpio-switch has uci-integration.

chunkeey commented on target/linux/mpc85xx/patches-4.14/104-powerpc-mpc85xx-change-P2020RDB-dts-file-for-OpenWRT.patch:
```diff
@@ -0,0 +1,128 @@
+ 	soc: soc@ffe00000 {
+ 		ranges = <0x0 0x0 0xffe00000 0x100000>;
+ 
++		gpio0: gpio-controller@fc00 {
```
and with gpio-switch this node can go...


It's a bit odd that you specify the PHY reset gpio when there's no network integration for this device.

CHKDSK88 replied:
@chunkeey 

>It's a bit odd that you specify the PHY reset gpio when there's no network integration for this device.

Device have 3 network controllers. Two of them have normal PHY. eth0 is connected to switch. At this moment device can be router, but without managed switch.

chunkeey replied:
So, I assume that the PHY reset gpio is the reset line for the switch?

chunkeey replied:
Thanks for clearing this up.

chunkeey commented on target/linux/mpc85xx/patches-4.14/104-powerpc-mpc85xx-change-P2020RDB-dts-file-for-OpenWRT.patch:
```diff
@@ -0,0 +1,128 @@
++		};
++
+ 		i2c@3000 {
++			adt7461@4c {
```
https://github.com/devicetree-org/devicetree-specification/blob/master/source/devicetree-basics.rst#generic-names-recommendation

maybe you could find a better node-name there.

chunkeey commented on target/linux/mpc85xx/patches-4.14/104-powerpc-mpc85xx-change-P2020RDB-dts-file-for-OpenWRT.patch:
```diff
@@ -0,0 +1,128 @@
+ 			flash@0 {
+ 				#address-cells = <1>;
+ 				#size-cells = <1>;
+-				compatible = "spansion,s25sl12801", "jedec,spi-nor";
```
Interesting. "spansion,s25sl12801" dt compat itself is not registered to any special chip and it seems to exclusively appear in other freescale boards (p1010,p102x, mpc8536, ...) .

However, it does not look like that you are using the spi-nor chip? So, you could just ignore this hunk. The [jedec,spi-nor.txt DT binding text](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/mtd/jedec%2Cspi-nor.txt) does not necessarily forbid anything other than "jedec,spi-nor".

    May include a device-specific string consisting of the
    manufacturer and name of the chip. A list of supported chip
    names follows.
    Must also include "jedec,spi-nor" for any SPI NOR flash that can
    be identified by the JEDEC READ ID opcode (0x9F)."

CHKDSK88 replied:
Ok. I drop my changes.

chunkeey commented on target/linux/mpc85xx/patches-4.14/104-powerpc-mpc85xx-change-P2020RDB-dts-file-for-OpenWRT.patch:
```diff
@@ -0,0 +1,128 @@
+ 
+ / {
+-	model = "fsl,P2020RDB";
++	model = "Freescale P2020RDB";
```
This is complicated. The current [DT Spec](https://github.com/devicetree-org/devicetree-specification/blob/211bbc01adb5bc5e64cb651651c4b907ed2d309f/source/devicetree-basics.rst#model) 



    Description:

        The model property value is a <string> that specifies the manufacturer’s model number of the device.

        The recommended format is: "manufacturer,model", where manufacturer is a string describing the name of the manufacturer (such as a stock ticker symbol), and model specifies the model number.

    Example:

        model = "fsl,MPC8349EMITX";

so :laughing: . 

Note:
I'm totally fine with "Freescale P2020RDB". This is just a FYI in case you want to upstream some of your changes.

CHKDSK88 replied:
Thank You for information.

mkresin commented on target/linux/mpc85xx/patches-4.14/104-powerpc-mpc85xx-change-P2020RDB-dts-file-for-OpenWRT.patch:
```diff
@@ -0,0 +1,128 @@
+ 		pci1 = &pci1;
+ 	};
+ 
++	gpio_export {
```
Assuming it isn't your intention to have the gpio available from userspace, [gpio-hog](https://www.kernel.org/doc/Documentation/devicetree/bindings/gpio/gpio.txt) is the upstream way to set GPIO early in the boot process to a defined state.

Beside of that, I would like to prevent adding our gpio-export hack to more targets. It was rejected upstream a long time ago and a lot of drivers are able to pull gpios on their own now days.

Maybe kernel commit bafbdd527d56 ("phylib: Add device reset GPIO support") does the job for you. A word of warning, I tried to use the new dts property on my own and failed back in the days (maybe it's fixed already):

  The per PHY reset GPIO code isn't able to bring a PHY out of reset,  which is initially in reset.
  
  The get_phy_device() in of_mdiobus_register_phy() returns an error, since the phy id for a PHY in reset can't be read (is 0xfffffff). The code deasserting the optional (initial) is called via phy_device_register() which is never reached in such a situation.

I have this issue on my todo list since april this year, but never managed to check if it can be easily fixed.

CHKDSK88 replied:
I add support for 4.19 kernel and use reset-gpio from upstream. 

So far everything work as expected.

CHKDSK88 commented:
My plan for this device:

0. Work with 4.19 kernel.
1. Try to add  eLBC support for [VSC7385 driver](https://github.com/torvalds/linux/blob/9a76aba02a37718242d7cdc294f0a3901928aa57/drivers/net/dsa/vitesse-vsc73xx.c).
2. Make VSC7385 managed.
2. Try to fix SD-card reader.

chunkeey commented:
> 0. Work with 4.19 kernel.

This can be a separate PR 

> 1. Try to add eLBC support for VSC7385 driver.
> 2. Make VSC7385 managed.

Hm, I didn't find any easily accessible datasheet or switch driver code. 
So far, I found a tiny firmware loader for the switch  
https://github.com/Freescale/u-boot-fslc/blob/2018.09%2Bfslc/drivers/net/vsc7385.c

which could help to map some of the existing spi-addresses to mmio.

Because looking from afar it seems like you would only need to implement the vsc73xx_read and vsc73xx_write in order to talk to the switch.  But I'm guessing.

CHKDSK88 commented:
> This can be a separate PR

Ok. I add separate PR for commit 84ccfcd8d0e

>Because looking from afar it seems like you would only need to implement the vsc73xx_read and vsc73xx_write in order to talk to the switch. But I'm guessing.

I hope so.

CHKDSK88 committed:
kernel/modules: add kmod-pmbus-zl6100 module

This patch adds the kmod packaging for the Intersil / Zilker Labs
ZL6100 and compatible digital DC-DC controllers as well as the
core kernel module for the Power Management Bus.

Add:
kmod-pmbus-core
kmod-pmbus-zl6100

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

chunkeey commented:
@CHKDSK88, so what to do with this PR then now that patch 1-3 have been merged and the last patch "mpc85xx: add support for Freescale (NXP) P2020RDB" requires 4.19? I can add it to my staging tree for the time being, but chances are that once we can do the switch something new will turn up. Do you want to keep it open (and do a rebase) or will you do another PR once the time comes?

(Note: I have to admit, due to the top-down nature of github and patchwork, new PRs get more attention.)

CHKDSK88 commented:
@chunkeey 

If You merge #1710 , I add one more kmod dependency to this PR.

I want to work with driver switch, but it will take some time. At this moment device could be used as regular router. It have two separate, full supported WANs and unmanaged LAN. I think, after merging #1710, it can be merged as initial support with added zl6100 module.

CHKDSK88 commented:
@chunkeey 

Sorry for misunderstanding. PHY can work without reset gpio. P2020RDB is reference board. I try to give maximum support of all peripherals.

4.19 is required only for work with VSC7385 DSA driver from upstream.

chunkeey commented:
Ok, one final question: @blocktrron wants to enable the ramdisk feature for the mpc85xx in his patchwork patch: https://patchwork.ozlabs.org/patch/1021672/ .

In your PR there's a ```#FEATURES += ramdisk```. So If I merge both patches the initramfs image will be build for the p2020 target then. Do you both think you both can come to a agreement there if the ramdisk should be available for the whole mpc85xx target or just the subtargets?

blocktrron commented:
@chunkeey It's fine for me as long it's enabled for the generic subtarget (as it's required for the Sophos installation).

CHKDSK88 commented:
@chunkeey  
Initramfs image for P2020RDB is working. 

But is problem for WDR4900.  Without any changes Image can't be uploaded: 

```
TFTP from server 192.168.0.66; our IP address is 192.168.0.86
Filename 'wdr4900v1_tp_recovery.bin'.
Load address: 0x1000000
Loading: T T T T T T T #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         ########################################
done
Bytes transferred = 4529439 (451d1f hex)
auto update firmware: product id verify fail!
```

I added to ``define Device/tl-wdr4900-v1``: 
```
  KERNEL_INITRAMFS := kernel-bin | tplink-v1-header
  KERNEL_ENTRY := 0x01100000
  KERNEL_LOADADDR := 0x01000000
```


But I get:

```
## Booting kernel from Legacy Image at 02000000 ...
   Image Name:   Linux-4.14.91
   Image Type:   PowerPC Linux Kernel Image (gzip compressed)
   Data Size:    4529123 Bytes = 4.3 MiB
   Load Address: 01000000
   Entry Point:  01000594
   Verifying Checksum ... Bad Data CRC
ERROR: can't get kernel image!
=>
```

At this moment initramfs image should be disabled for WDR4900.

chunkeey commented:
> At this moment initramfs image should be disabled for WDR4900.

the ramdisk feature is a per target/sub-target setting. I don't think it's possible to disable it for just one device of said target. (which is bad because @blocktrron needs ramdisk for the generic target I think)

> ```  KERNEL_INITRAMFS := kernel-bin | tplink-v1-header ```

Didn't we remove "tplink-v1-header" in favor of mktplinkfw? Did you add it back?

Usually, initramfs kernels shouldn't need any special headers and looking at the WDR4900 I think just 
 ```  KERNEL_INITRAMFS := kernel-bin ``` could be enough (which should be the default anyway).
But I don't have a WDR4900 so, I can only guess.

As for:
>     Verifying Checksum ... Bad Data
>     CRC ERROR: can't get kernel image!

Could be a fluke. Have you checked with the [crc32 u-boot](https://www.denx.de/wiki/DULG/UBootCmdGroupMemory) command if the uploaded image was not corrupted? The u-boot
network drivers and stack do have bugs as well (especially really old ones that never get
any updated).

CHKDSK88 commented:
>Didn't we remove "tplink-v1-header" in favor of mktplinkfw? Did you add it back?

I tried use generic "tplink-v1-header", which is used for initramfs images.

> Usually, initramfs kernels shouldn't need any special headers and looking at the WDR4900 I think just
KERNEL_INITRAMFS := kernel-bin could be enough (which should be the default anyway).
But I don't have a WDR4900 so, I can only guess.

It isn't. It causes:

```auto update firmware: product id verify fail!```

>Could be a fluke. Have you checked with the crc32 u-boot command if the uploaded image was not corrupted? The u-boot
network drivers and stack do have bugs as well (especially really old ones that never get
any updated).

I have WDR4900 with blocked u-boot. Only recovery works. I tried two times.


My proposal:

```
diff --git a/target/linux/mpc85xx/image/Makefile b/target/linux/mpc85xx/image/Makefile
index 1b4b0ecf33..e80df25072 100644
--- a/target/linux/mpc85xx/image/Makefile
+++ b/target/linux/mpc85xx/image/Makefile
@@ -49,6 +49,7 @@ define Device/tl-wdr4900-v1
   TPLINK_HWREV := 1
   TPLINK_FLASHLAYOUT := 16Mppc
   KERNEL_NAME := cuImage.tl-wdr4900-v1
+  KERNEL_INITRAMFS :=
   SUPPORTED_DEVICES:=tl-wdr4900-v1 tplink,tl-wdr4900-v1
   IMAGES := fdt.bin factory.bin sysupgrade.bin
   IMAGE/fdt.bin := append-dtb
```

This workaround causes, that initramfs image isn't generated for this device. I think, it's better solution than not working image.

chunkeey commented:
> I have WDR4900 with blocked u-boot. Only recovery works. I tried two times.

Ok, thanks this was crucial information.

> My proposal:
>
>```
>diff --git a/target/linux/mpc85xx/image/Makefile b/target/linux/mpc85xx/image/Makefile
>index 1b4b0ecf33..e80df25072 100644
>--- a/target/linux/mpc85xx/image/Makefile
>+++ b/target/linux/mpc85xx/image/Makefile
>@@ -49,6 +49,7 @@ define Device/tl-wdr4900-v1
>    TPLINK_HWREV := 1
>    TPLINK_FLASHLAYOUT := 16Mppc
>    KERNEL_NAME := cuImage.tl-wdr4900-v1
>+  KERNEL_INITRAMFS :=
>    SUPPORTED_DEVICES:=tl-wdr4900-v1 tplink,tl-wdr4900-v1
>    IMAGES := fdt.bin factory.bin sysupgrade.bin
>    IMAGE/fdt.bin := append-dtb
>```
>This workaround causes, that initramfs image isn't generated for this device. I think, it's better solution than not working image.

TIL. Thanks.

Can I get some sort of commit message and signed-off-by so I can add it to the mpc85xx?

CHKDSK88 committed:
mpc85xx: disable initramfs image in TL-WDR4900

Initramfs image isn't required for this device and regular
initramfs generation isn't work properly. It create not working
binaries.

This patch disable initramfs image for TL-WDR4900.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 commented:
@chunkeey 

I added two commits: first from  #1710, second created from [this comment.](https://github.com/openwrt/openwrt/pull/1688#issuecomment-453742502)

I cleaned commented ramdisk feature and add zl6100 module to p2020rdb.

I think, its't ready to merge.

chunkeey commented:
Merged. Thanks.

chunkeey commented:
https://github.com/openwrt/openwrt/commit/c7102a76992d9021aa6abab7b7c0e7758f765668

at least it got caught pretty quickly.
